pkg/controller/context.go

```diff
@@ -167,3 +209,136 @@ type SchedulerOptions struct {
 	// scheduled as 'processing' at once.
 	MaxConcurrentChallenges int
 }
+
+// ContextFactory is used for constructing new Contexts who's clients have been
+// configured with a User Agent built from the component name.
+type ContextFactory struct {
+	// baseRestConfig is the base Kubernetes REST config that can authenticate to
+	// the Kubernetes API server.
+	baseRestConfig *rest.Config
+
+	// log is the factory logger which is used to construct event broadcasters.
+	log logr.Logger
+
+	// ctx is the base controller Context that all Contexts will be built from.
+	ctx *Context
+}
+
+// NewContextFactory builds a ContextFactory that builds controller Contexts
+// that have been configured for that components User Agent.
+func NewContextFactory(ctx context.Context, opts ContextOptions) (*ContextFactory, error) {
+	// Load the users Kubernetes config
+	restConfig, err := clientcmd.BuildConfigFromFlags(opts.APIServerHost, opts.Kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating rest config: %w", err)
+	}
+	restConfig = util.RestConfigWithUserAgent(restConfig)
+
+	clients, err := buildClients(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	sharedInformerFactory := informers.NewSharedInformerFactoryWithOptions(clients.cmClient, resyncPeriod, informers.WithNamespace(opts.Namespace))
+	kubeSharedInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(clients.kubeClient, resyncPeriod, kubeinformers.WithNamespace(opts.Namespace))
+	gwSharedInformerFactory := gwinformers.NewSharedInformerFactoryWithOptions(clients.gwClient, resyncPeriod, gwinformers.WithNamespace(opts.Namespace))
+
+	return &ContextFactory{
+		baseRestConfig: restConfig,
+		log:            logf.FromContext(ctx),
+		ctx: &Context{
+			RootContext:               ctx,
+			StopCh:                    ctx.Done(),
+			KubeSharedInformerFactory: kubeSharedInformerFactory,
+			SharedInformerFactory:     sharedInformerFactory,
+			GWShared:                  gwSharedInformerFactory,
+			GatewaySolverEnabled:      clients.gatewayAvailable,
+			ContextOptions:            opts,
+		},
+	}, nil
+}
+
+// Build builds a new controller Context who's clients have a User Agent
+// derived from the optional component name.
+func (c *ContextFactory) Build(component ...string) (*Context, error) {
+	restConfig := util.RestConfigWithUserAgent(c.baseRestConfig, component...)
+
+	clients, err := buildClients(restConfig)
```

Comment:
Thanks, I had missed these two values. I have updated the PR to assign these values inside the ContextFactory constructor. All clients will therefore inherit these vaules.

Reply:
Thanks @JoshVanL , but what I meant was - is it not the case that with this PR because we are now constructing a new client for each component, we also are setting up a separate rate limit bucket for each so the qps and burst is no longer the limit of cert-manager controller queries, but the limit of each separate component's queries?

Reply:
i.e, will we not be now creating a new [`RateLimiter`](https://github.com/kubernetes/client-go/blob/v0.23.3/kubernetes/clientset.go#L431-L435) once for each controller?

Reply:
Thanks @irbekrm, you're right in the current implementation it will be creating a new set of buckets for each Factory that gets built.

Looking at that code path it looks like we can set our own `RateLimiter` which won't get overridden, and which we can preserve across Contexts. This looks like a good resolution- will update the PR